# optDNTRA: Optimization of De Novo Transcriptome RNA-seq Assembly

## Overview

`optDNTRA` is a command-line tool tailored to optimize transcript assemblies generated by the RNA-seq assembler (e.g. Trinity). It simplifies data processing for both single-end and paired-end RNA-seq datasets and offers a range of quality control and assessment options. The tool is fully integrated with Snakemake, ensuring reproducible and scalable workflows.

## Installation

To install the required dependencies for `optDNTRA`, we recommend using [Mamba](https://github.com/mamba-org/mamba), a fast, robust package manager that is a drop-in replacement for Conda. Mamba significantly speeds up the installation process and handles complex dependency resolution more efficiently.

To set up `optDNTRA`, first ensure you have Mamba installed. Then, clone the repository and create an environment as follows:

```bash
git clone https://github.com/zywu2002/optDNTRA.git
cd optDNTRA
mamba env create -f environment.yml
mamba activate optDNTRA
```

## Usage

Here's a summary of the available command-line arguments:

```bash
$ python optDNTRA.py -h

usage: optDNTRA.py [-h] -t transcripts.fasta [-f reads.fq] [--left reads_1.fq] [--right reads_2.fq] [--sampleSheet samples.tab]
                   [--reference reference.fasta] [--singleEnd] [--ss-lib-type {F,R,RF,FR}] [--trim] [--qc] [--buscoAsmt] [--omarkAsmt]
                   [--emapper] [-v] [--threads THREADS] [--snakemakeOptions SNAKEMAKEOPTIONS]

 
optDNTRA: optimization of De Novo Transcriptome Rna-seq Assembly

For RNA-seq input data:
    If Paired-end:
        --left <string>         left reads
        --right <string>        right reads
    or Single-end:
        --fastq, -f <string>    single-end reads
    or
        --sampleSheet <string>  tab-delimited text file indicating biological replicate relationships
                                e.g.
                                cond_A    cond_A_rep1    A_rep1_left.fq    A_rep1_right.fq
                                cond_A    cond_A_rep2    A_rep2_left.fq    A_rep2_right.fq
                                cond_B    cond_B_rep1    B_rep1_left.fq    B_rep1_right.fq
                                cond_B    cond_B_rep2    B_rep2_left.fq    B_rep2_right.fq
                                if single-end, remove the 4th column in the text file.
        

options:
  -h, --help            show this help message and exit
  -t transcripts.fasta, --transcript transcripts.fasta
                        transcript fasta file
  -f reads.fq, --fastq reads.fq
                        single-end reads
  --left reads_1.fq     left reads
  --right reads_2.fq    right reads
  --sampleSheet samples.tab
                        tab-delimited text file indicating biological replicate relationships
  --reference reference.cdna.fasta
                        reference transcriptome
  --singleEnd           specify if the input data is single-end, not paired-end
  --ss-lib-type {F,R,RF,FR}
                        strand-specific library type: single('F' or 'R'), paired('RF' or 'FR')
  --trim                enable trimming for input data
  --qc                  enable quality control for input data
  --buscoAsmt           enable BUSCO assessment
  --omarkAsmt           enable OMArk assessment
  --emapper             enable EggNOG-mapper for functional annotation
  -v, --verbose         print detailed reports
  --threads THREADS     number of threads to use
  --snakemakeOptions SNAKEMAKEOPTIONS
                        Snakemake options to be passed directly to snakemake, e.g. use --snakemakeOptions='--dryrun'
```

## Example Command

For a practical usage scenario, consider the following command to run optDNTRA on paired-end reads, enabling trimming and quality control:

```bash
python optDNTRA.py -t transcripts.fasta --left left_reads.fq --right right_reads.fq --threads 8 --trim --qc
```

## Output

Upon completion, `optDNTRA` generates several output files in the specified directory. The key output files include:

- `results/02-optimization/03-transEvidence/transcript.flt.final.fa` - Optimized transcript assemblies with low-quality or redundant sequences removed, providing a refined set of transcripts for downstream analysis.

- `results/02-optimization/03-transEvidence/transcript.flt.final.pep` - Proteins fasta file from optimized transcript assemblies.

- `results/03-assessment/Busco/` (if enabled) - Includes completeness scores of the transcriptome based on Benchmarking Universal Single-Copy Orthologs.

- `results/03-assessment/Omark/` (if enabled) - Details the functional assessment results.

- `results/04-annotation/transAsm.emapper.annotations` (if enabled) - EggNOG-mapper functional annotations mapped to transcripts.

Each of these files is organized in a results folder, ensuring easy access to quality control metrics, optimized transcripts, and assessment data for further analysis or reporting.
